Mathlib/RingTheory/DividedPowers/DPMorphism.lean

```diff
@@ -0,0 +1,242 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir, María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández
+-/
+import Mathlib.RingTheory.DividedPowers.Basic
+
+/-! # Divided power morphisms
+
+Let `A` and `B` be commutative (semi)rings, let `I` be an ideal of `A` and let `J` be an ideal of
+`B`. Given divided power structures on `I` and `J`, a ring morphism `A →+* B` is a *divided
+power morphism* if it is compatible with these divided power structures.
+
+## Main definitions
+
+* `DividedPowers.IsDPMorphism` : given divided power structures on the `A`-ideal `I` and the
+  `B`-ideal `J`, a ring morphism `A →+* B` is a divided power morphism if it is compatible with
+  these divided power structures.
+* `DividedPowers.DPMorphism` : a bundled version of `IsDPMorphism`.
+* `DividedPowers.ideal_from_ringHom` : given a ring homomorphism `A →+* B` and ideals `I ⊆ A` and
+  `J ⊆ B` such that `I.map f ≤ J`, this is the `A`-ideal on which
+  `f (hI.dpow n x) = hJ.dpow n (f x)`.
+* `DividedPowers.DPMorphism.fromGens` : the `DPMorphism` induced by a ring morphism, given that
+  divided powers are compatible on a generating set.
+
+## Main results
+
+* `DividedPowers.dpow_eq_from_gens` : if two divided power structures on an ideal `I` agree on a
+  generating set, then they are equal.
+
+
+## References
+
+* [P. Berthelot (1974), *Cohomologie cristalline des schémas de
+caractéristique $p$ > 0*][Berthelot-1974]
+
+* [P. Berthelot and A. Ogus (1978), *Notes on crystalline
+cohomology*][BerthelotOgus-1978]
+
+* [N. Roby (1963), *Lois polynomes et lois formelles en théorie des
+modules*][Roby-1963]
+
+* [N. Roby (1965), *Les algèbres à puissances dividées*][Roby-1965]
+-/
+
+open Ideal Set SetLike
+
+namespace DividedPowers
+
+/-- Given divided power structures on the `A`-ideal `I` and the `B`-ideal `J`, a ring morphism
+  `A → B` is a divided power morphism if it is compatible with these divided power structures. -/
+def IsDPMorphism {A B : Type*} [CommSemiring A] [CommSemiring B] {I : Ideal A} {J : Ideal B}
+    (hI : DividedPowers I) (hJ : DividedPowers J) (f : A →+* B) : Prop :=
+  I.map f ≤ J ∧ ∀ {n : ℕ}, ∀ a ∈ I, hJ.dpow n (f a) = f (hI.dpow n a)
```

Comment:
What about making this into a structure with two fields? It seems more natural (and mathematically it is exactly the same).